apisix/discovery/k8s.lua

```diff
@@ -0,0 +1,472 @@
+        return false, "UnexpectedBody", body
+    end
+
+    resource.newest_resource_version = data.metadata.resourceVersion
+
+    for _, item in ipairs(data.items or empty_table) do
+        resource:event_dispatch("ADDED", item)
+    end
+
+    if data.metadata.continue ~= nil and data.metadata.continue ~= "" then
+        list_resource(httpc, resource, data.metadata.continue)
+    end
+
+    return true, "Success", ""
+end
+
+local function watch_resource(httpc, resource)
+    math.randomseed(process.get_master_pid())
+    local watch_seconds = 1800 + math.random(60, 1200)
+    local allowance_seconds = 120
+    httpc:set_timeouts(2000, 3000, (watch_seconds + allowance_seconds) * 1000)
+    local res, err = httpc:request({
+        path = resource:watch_path(),
+        query = resource:watch_query(watch_seconds),
+        headers = {
+            ["Host"] = string.format("%s:%s", apiserver_host, apiserver_port),
+            ["Authorization"] = string.format("Bearer %s", apiserver_token),
+            ["Accept"] = "application/json",
+            ["Connection"] = "keep-alive"
+        }
+    })
+
+    if err then
+        return false, "RequestError", err
+    end
+
+    if res.status ~= 200 then
+        return false, res.reason, res:read_body() or ""
+    end
+
+    local remainder_body = ""
+    local body = ""
+    local reader = res.body_reader
+    local gmatch_iterator;
+    local captures;
+    local captured_size = 0
+    while true do
+
+        body, err = reader()
+        if err then
+            return false, "ReadBodyError", err
+        end
+
+        if not body then
+            break
+        end
+
+        if #remainder_body ~= 0 then
+            body = remainder_body .. body
+        end
+
+        gmatch_iterator, err = ngx.re.gmatch(body, "{\"type\":.*}\n", "jiao")
+        if not gmatch_iterator then
+            return false, "GmatchError", err
+        end
+
+        while true do
+            captures, err = gmatch_iterator()
+            if err then
+                return false, "GmatchError", err
+            end
+            if not captures then
+                break
+            end
+            captured_size = captured_size + #captures[0]
+            local v, _ = core.json.decode(captures[0])
+            if not v or not v.object or v.object.kind ~= resource.kind then
+                return false, "UnexpectedBody", captures[0]
+            end
+
+            resource.newest_resource_version = v.object.metadata.resource_version
+            if v.type ~= "BOOKMARK" then
+                resource:event_dispatch(v.type, v.object)
+            end
+        end
+
+        if captured_size == #body then
+            remainder_body = ""
+        elseif captured_size == 0 then
+            remainder_body = body
+        else
+            remainder_body = string.sub(body, captured_size + 1)
+        end
+    end
+    watch_resource(httpc, resource)
+end
+
+local function fetch_resource(resource)
```

Comment:
It seems that this function will be run inside a timer for a long while, this may cause the temporary memory leaky, we may use a counter to force it to quit and use another timer, see the below for details.

https://github.com/apache/apisix/blob/master/apisix/core/config_etcd.lua#L534